### PR TITLE
Auto grade assessment and auto release grade

### PR DESCRIPTION
**Pull request by allenwq**
Depends on #1224 

If assessment is auto graded, grade the assessment and release the grades and exp.

### CODE REVIEW

File: config/locales/en/course/assessment/submission/submissions.yml [hunk 1]
```diff
@@ -31,6 +31,7 @@ en:
             grade_summary: 'Grade Summary'
           update:
             success: 'Submission was updated.'
+            finalise: 'Your submission will be auto graded, please come to this page later.'
```
Comment (kxmbrian) [inline config/locales/en/course/assessment/submission/submissions.yml]: Maybe, "Your submission is being graded, please visit this page again later."

File: app/services/course/assessment/submission/auto_grading_service.rb [hunk 2]
```diff
@@ -65,4 +66,22 @@ def aggregate_failures(jobs)
...
+  # Calculating scheme:
+  #   Submit in time: ( base_exp + bonus_exp ) * actual_grade / max_grade
+  #   Late submission: base_exp * actual_grade / max_grade
+  def calculate_exp(submission)
+    assessment = submission.assessment
+    total_exp = assessment.base_exp
+    if !assessment.bonus_end_at || submission.submitted_at <= assessment.bonus_end_at
+      total_exp += assessment.time_bonus_exp
+    end
+    (submission.grade.to_f / assessment.maximum_grade * total_exp).to_i
```
Comment (kxmbrian) [inline app/services/course/assessment/submission/auto_grading_service.rb]: We can currently set questions to have a maximum grade of 0. Should we return full exp if assessment max grade is zero?

Comment (allenwq) [inline app/services/course/assessment/submission/auto_grading_service.rb]: Yes, I should handle that, thanks !

File: app/services/course/assessment/submission/auto_grading_service.rb [hunk 3]
```diff
@@ -65,4 +66,25 @@ def aggregate_failures(jobs)
...
+
+  def publish_grade(submission)
+    submission.answers.select(&:submitted?).each(&:publish!)
+    submission.points_awarded = calculate_exp(submission).to_i
+    submission.publish!
+  end
+
+  # Calculating scheme:
+  #   Submit in time: ( base_exp + bonus_exp ) * actual_grade / max_grade
+  #   Late submission: base_exp * actual_grade / max_grade
```
Comment (fonglh) [inline app/services/course/assessment/submission/auto_grading_service.rb]: This isn't actually late right? It's a submission after the bonus cutoff time?

Comment (allenwq) [inline app/services/course/assessment/submission/auto_grading_service.rb]: Yes it's after bonus cutoff, so I described it as late, any better names ?

Comment (fonglh) [inline app/services/course/assessment/submission/auto_grading_service.rb]: Late implies that it's after the whole assessment's due date, not just the bonus. How about
"Submit before bonus cutoff" and "Submit after bonus cutoff" ?

**Pull request merged**